### PR TITLE
Fix #43664 pip module is not working under windows.

### LINKED ISSUE #43664: pip module is not working under windows
PIP salt module is not working under windows environment.
The problem is that during salt-minion installer creation there is cleanup of the absolute path to python.exe executable this is leading to `Fatal error in launcher: Unable to create process using '"' 
` error when pip is executed. Better solution is to replace c:\python{{version}} with c:\salt\bin path which will allows to successfully execute pip and install additional modules (in my case it was dnspython).
So line as the one below (and for pip2, pip27 exe files) https://github.com/saltstack/salt/blob/8cf0588221fec3eddf310bff09c9affbd17cf19b/pkg/windows/build_pkg.bat#L144

should become:
`"%PyDir%\python" "%CurDir%\portable.py" -f "%BinDir%\Scripts\pip.exe"  -r "c:\salt\bin"`.

### PR DESCRIPTION
**Pull request by voyvodov**
### What does this PR do?
Fix a bug in salt-minion installer for windows
### What issues does this PR fix or reference?
Fix #43664 issue.
### Previous Behavior
When salt minion is trying to execute pip module it fails. So pip.installed module cannot be used under Windows.

### New Behavior
Salt minion now can execute pip correctly, so new Python modules can be installed/upgraded/removed under Windows

### Tests written?

No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

### CODE REVIEW

**Comment by skizunov** [comment]
I am not a fan of hard-coded paths on Windows because not everyone is going to install to `C:\salt\bin`. Some users may opt to install to a different disk drive. Others may have a different default install path. Is there another way to make this work?

**Comment by hristozv** [comment]
Yeah, me too, however c:\salt path is hardcoded in Salt-Minion-Setup.nsi file already and installer doesn't provide way to change it.
Still, for me, better solution is to generally change nsi file to provide installation directory option and execute portable.py (or something similar) script during the installation.

**Comment by skizunov** [comment]
My usage of Salt is a little bit different, and I provide my own installer which installs to a different path.

**Comment by skizunov** [comment]
And that path is user configurable in the installer that I provide. This is why hard-coding, no matter what the value, would break my use case, since the final path is unknown until install time.

**Comment by hristozv** [comment]
What about if portable.py is used once again during the installation to replace those `..\` with the correct $INSTDIR and be deleted after that? Does it sounds better?
If yes, please close this PR, I will open another once with such fix.

**Comment by skizunov** [comment]
I tested this, and it seems that pip works as it was before (with the relative path) if the current directory is the scripts directory. It fails if the current directory is not the scripts directory though. Could the pip module be modified to make the calls into pip with the current directory set? `cmd.run` and `cmd.run_all` do have a `cwd` parameter that could be leveraged to achieve this. This would be in Windows-only logic, of course.

Thoughts on this approach?

**Comment by hristozv** [comment]
I think it can be done, but still that means to hardcode some path in the states. So just moving one problem from the installer to the states. I still prefer set hardcoded path once the installer completes on which I'm working now. That way it solves all the issues with single change.

**Comment by skizunov** [comment]
No hard-coding required: Just use:
```
os.path.join(__opts__['root_dir'], 'bin', 'Scripts')
```
This type of change would be my preferred solution.

**Review by twangboy — COMMENTED** [review]
We shouldn't be hard-coding paths here

**Comment by twangboy** [comment]
I've always thought the pip module should default to the salt installation pip.

**Comment by twangboy** [comment]
Or we change how the installer works.
- Package portable.py with the installer
- Run it on install with the path to the salt installation

That might be a better solution as you would only have to pass `bin_env` with pip module instead of both `bin_env` and `cwd`

**Comment by twangboy** [comment]
@UtahDave ^^^

**Comment by hristozv** [comment]
Yeap, exactly what I was thinking. I think this PR can be closed and I will create a new once I update the installer to set the path based on the $INSTDIR value from the installer. This way it will work without messing with the module or set cwd/bin path arguments.

**Comment by skizunov** [comment]
Yes, my idea has a flaw in it now that I am looking at the pip module. It seems like the user can specify their own cwd for `pip.install`. If we also need to specify the cwd (based on it only running correctly from the `Scripts` subdir as the cwd), then there is a conflict since we can't have two cwd's at once.

**Comment by hristozv** [comment]
Now I revert the changes in build_pkg.bat file. Instead the nsi installer script is changed to replace path to the correct one in pip.exe file.
Probably is better to remove all replacement in build script and make them in the installer.

**Comment by twangboy** [comment]
I've been working on the installer... I'm waiting for a PR to be merged. I can take a look at this, so it doesn't conflict with my recent changes.

**Comment by cachedout** [comment]
@twangboy What's the current status here? Has the PR that you were waiting on been merged?

**Comment by twangboy** [comment]
@cachedout Still waiting: https://github.com/saltstack/salt/pull/43649

**Comment by cachedout** [comment]
@twangboy The issue you reference (#43649) has now been merged. Could you please come back to this? Thanks.

File: pkg/windows/installer/Salt-Minion-Setup.nsi [hunk 1]
```diff
@@ -285,6 +285,16 @@ Section "MainSection" SEC01
     File /r "..\buildenv\"
     nsExec::Exec 'icacls c:\salt /inheritance:r /grant:r "*S-1-5-32-544":(OI)(CI)F /grant:r "*S-1-5-18":(OI)(CI)F'
 
+    ; Extract portable.py script so we can fix paths in pip.exe
+    SetOutPath "$TEMP\"
+    File "..\portable.py"
+
+    ; Update python pip module with correct path.
+    nsExec::Exec '$INSTDIR\bin\python.exe "$TEMP\portable.py" -f "$INSTDIR\bin\Scripts\pip.exe" -s "..\\" -r $INSTDIR\bin\'
```
Comment (skizunov) [inline pkg/windows/installer/Salt-Minion-Setup.nsi]: `..\` seems too general for a search and replace in an entire executable file.

What about either:
```
-s "..\\python" -r $INSTDIR\bin\python'
```
or
```
-s "..\\python.exe" -r $INSTDIR\bin\python.exe'
```
just to be a bit safer. The second one will only work on 64-bit installs, I believe, due to the name of the exe on 32-bit being different. The first one should work on both (but is slightly less safe).

Thoughts?

Comment (hristozv) [reply pkg/windows/installer/Salt-Minion-Setup.nsi]: Look fine for me. The first one seems better to avoid 32/64 bits differences.

Comment (skizunov) [reply pkg/windows/installer/Salt-Minion-Setup.nsi]: I am fine with using the first one then.

Comment (twangboy) [reply pkg/windows/installer/Salt-Minion-Setup.nsi]: You could probably remove the entire section of `build_pkg.bat` where it makes those portable:
```
:: Remove the fixed path in .exe files
@echo Removing fixed path from .exe files
@echo ----------------------------------------------------------------------
"%PyDir%\python" "%CurDir%\portable.py" -f "%BinDir%\Scripts\easy_install.exe"
"%PyDir%\python" "%CurDir%\portable.py" -f "%BinDir%\Scripts\easy_install-%PyVerMajor%.%PyVerMinor%.exe"
"%PyDir%\python" "%CurDir%\portable.py" -f "%BinDir%\Scripts\pip.exe"
"%PyDir%\python" "%CurDir%\portable.py" -f "%BinDir%\Scripts\pip%PyVerMajor%.%PyVerMinor%.exe"
"%PyDir%\python" "%CurDir%\portable.py" -f "%BinDir%\Scripts\pip%PyVerMajor%.exe"
"%PyDir%\python" "%CurDir%\portable.py" -f "%BinDir%\Scripts\wheel.exe"
@echo.
```

and then modify them in the nullsoft script without specifying a search term:
```
nsExec::Exec '$INSTDIR\bin\python.exe "$TEMP\portable.py" -f "$INSTDIR\bin\Scripts\pip.exe" -r $INSTDIR\bin\'
```

Need to figure out a way to get Major and Minor versions of Python

**Comment by rallytime** [comment]
@hristozv Just checking in here, but did you see the comments above from @twangboy?

**Comment by twangboy** [comment]
This needs to be closed. This will be fixed when we do this: https://github.com/saltstack/salt/issues/42801

**Comment by twangboy** [comment]
@rallytime

**Comment by rallytime** [comment]
Since we haven't had a response here and due to the discussion above, I am going to close this PR. If the questions above can be addressed, we can of course re-open this. Or a new PR can be opened as well.

**Pull request closed without merging**